### ISSUE TITLE
Pyright reports error when type is any subtype of `pydantic.BaseModel`, ie. `type[pydantic.BaseModel]`

### ISSUE THREAD

**Issue by dingobar** [microsoft/pyright#4721]
Pyright reports an error in a function with an argument of type `type[pydantic.BaseModel]`, if a class that implements that class is passed to the function.

```py
class MyType(pydantic.BaseModel):
    my_field: int

def my_func(obj: dict[str, Any], input_type: type[MyType]) -> MyType:
    return input_type.parse_obj(obj)
```

This reports:
```
error: Expression of type "BaseModel" cannot be assigned to declared type "MyType"
```

Any ideas how to make pyright realize that `MyType` is a subtype of `BaseModel` and thus should be allowed here?

Python 3.10, MacOS, pyright running on CLI.

**Comment by dingobar** [comment]
Solved by using `TypeVar`:

```python
T = TypeVar("T", bound=BaseModel)

class MyType(pydantic.BaseModel):
    my_field: int

def my_func(obj: dict[str, Any], input_type: type[T]) -> T:
    return input_type.parse_obj(obj)

my_object: MyType = my_func({"my_field": 1}, MyType) 
```

However, I would love to understand why this is OK and the original example not?
